src/System.Net.HttpListener/tests/HttpRequestStreamTests.cs

```diff
@@ -0,0 +1,454 @@
+                Assert.Throws<ArgumentException>("asyncResult", () => inputStream2.EndRead(beginReadResult));
+            }
+        }
+
+        [Fact]
+        public async Task EndRead_CalledTwice_ThrowsInvalidOperationException()
+        {
+            HttpListenerRequest request = await _helper.GetRequest();
+            using (Stream inputStream = request.InputStream)
+            {
+                IAsyncResult beginReadResult = inputStream.BeginRead(new byte[0], 0, 0, new AsyncCallback(Common.StubCallback), null);
+                inputStream.EndRead(beginReadResult);
+
+                Assert.Throws<InvalidOperationException>(() => inputStream.EndRead(beginReadResult));
+            }
+        }
+
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public async Task Read_FromClosedSocketSynchronously_ThrowsHttpListenerException(bool transferEncodingChunked)
+        {
+            const string Text = "Some-String";
+            byte[] expected = Encoding.UTF8.GetBytes(Text);
+            Task<HttpListenerContext> contextTask = _listener.GetContextAsync();
+
+            using (HttpClient client = new HttpClient())
+            {
+                client.DefaultRequestHeaders.TransferEncodingChunked = transferEncodingChunked;
+                Task<HttpResponseMessage> clientTask = client.PostAsync(_factory.ListeningUrl, new StringContent(Text));
+
+                HttpListenerContext context = await contextTask;
+
+                // Disconnect the client from the HttpListener.
+                client.Dispose();
+
+                // Reading from a closed client should fail.
+                byte[] buffer = new byte[expected.Length];
+                Assert.Throws<HttpListenerException>(() => context.Request.InputStream.Read(buffer, 0, buffer.Length));
+
+                // Closing a response from a closed client should fail.
+                Assert.Throws<HttpListenerException>(() => context.Response.Close());
+            }
+        }
+
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public async Task Read_FromClosedSocketASynchronously_ThrowsHttpListenerException(bool transferEncodingChunked)
+        {
+            const string Text = "Some-String";
+            byte[] expected = Encoding.UTF8.GetBytes(Text);
+            Task<HttpListenerContext> contextTask = _listener.GetContextAsync();
+
+            using (HttpClient client = new HttpClient())
+            {
+                client.DefaultRequestHeaders.TransferEncodingChunked = transferEncodingChunked;
+                Task<HttpResponseMessage> clientTask = client.PostAsync(_factory.ListeningUrl, new StringContent(Text));
+
+                HttpListenerContext context = await contextTask;
+
+                // Disconnect the client from the HttpListener.
+                client.Dispose();
+
+                // TODO: we need to sleep and wait to make sure the client closes.
+                Thread.Sleep(1000);
```

Comment:
I'm not sure the best way to fix this TODO. I'm trying to test a scenario where we have connected to the client/socket, but fail to read from or write to it. The same goes for the other TODOs.

Without the thread sleep, these tests fail intermittently.